gencred/cmd/gencred/main.go

```diff
@@ -174,21 +272,132 @@ func Main() {
+	if o.refreshInterval == 0 {
+		if err := runOnce(c); err != nil {
+			util.PrintErrAndExit(err)
+		}
+		return
 	}
 
-	if err = writeConfig(o, clientset); err != nil {
-		util.PrintErrAndExit(err)
+	defer interrupts.WaitForGracefulShutdown()
+	interrupts.Tick(func() {
+		runOnce(c)
+	}, func() time.Duration { return o.refreshInterval })
+}
+
+func runOnce(c config) error {
+	// Make sure process everyone before crying.
+	var errs []error
+	for _, cc := range c.Clusters {
+		if cc.GKEConnection != nil && cc.Context != nil {
+			errs = append(errs, errors.New("gke and context are mutually exclusive"))
+			continue
+		}
+		if cc.DurationInSeconds == 0 {
+			cc.DurationInSeconds = defaultDurationInSeconds
+		}
+		var clientset *kubernetes.Clientset
+		if cc.Context != nil {
+			loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+			configOverrides := &clientcmd.ConfigOverrides{CurrentContext: *cc.Context}
+			kubeconfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides)
+
+			config, err := kubeconfig.ClientConfig()
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+
+			clientset, err = kubernetes.NewForConfig(config)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+		} else {
+			gkeService, err := container.NewService(context.Background())
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			cluster, err := container.NewProjectsLocationsClustersService(gkeService).Get(*cc.GKEConnection).Do()
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+
+			decodedClientCertificate, err := base64.StdEncoding.DecodeString(cluster.MasterAuth.ClientCertificate)
+			if err != nil {
+				fmt.Println("decode client certificate error:", err)
+				continue
+			}
+			decodedClientKey, err := base64.StdEncoding.DecodeString(cluster.MasterAuth.ClientKey)
+			if err != nil {
+				fmt.Println("decode client key error:", err)
+				continue
+			}
+			decodedClusterCaCertificate, err := base64.StdEncoding.DecodeString(cluster.MasterAuth.ClusterCaCertificate)
+			if err != nil {
+				fmt.Println("decode cluster CA certificate error:", err)
+				continue
+			}
+
+			config := &rest.Config{
+				Host: "https://" + cluster.Endpoint,
+				TLSClientConfig: rest.TLSClientConfig{
+					Insecure: false,
+					CertData: decodedClientCertificate,
+					KeyData:  decodedClientKey,
+					CAData:   decodedClusterCaCertificate,
+				},
+			}
+
+			cred, err := google.DefaultTokenSource(context.Background(), container.CloudPlatformScope)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			config.Wrap(func(rt http.RoundTripper) http.RoundTripper {
+				return &oauth2.Transport{
+					Source: cred,
+					Base:   rt,
+				}
+			})
+
+			clientset, err = kubernetes.NewForConfig(config)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("failed to initialise clientset from config: %s", err))
+				continue
+			}
```

Comment:
nit: This part is similar for Context or GKEConnection, could be moved out of the conditional.

Reply:
doen